thelper/optim/utils.py

```diff
@@ -181,13 +181,21 @@ def create_optimizer(config, model):
     via a dictionary named 'params'.
     """
     logger.debug("loading optimizer")
+    if isinstance(config, torch.optim.Optimizer):
+        # user passed in a fully instantiated optimizer; trust them and return it directly...
+        return config
```

Comment:
I think this can cause problem if the model loaded from the config was built after the optimizer passed as input from python code. The linked model-parameters wouldn't be the same.

Reply:
Wouldn't the optimizer always need the model's weights as input to its constructor, making it impossible to build the model after building the optimizer? (unless it's a different model, and then we can blame the user for the bad logic of trying to re-use an old optimizer?)

Reply:
Yes for building the optimizer, the model must be created first. 
I'm not sure though if the model would be reloaded or not when parsing the config.
The actual model instance created to make the optimizer wouldn't be used by thelper.

Reply:
The model should be instantiated only once before the instantiation of the optimizer, which is the responsibility of the session/trainer object (by default). This is enforced by the fact that model and trainer configs are separated and are defined at the same hierarchical level in the config files. In regular CLI use cases, `thelper.train.create_trainer` receives the fully instantiated model and passes it by reference to the trainer object's constructor. The trainer objects should never be calling the "create_model" function, or modifying model weights directly (except through using the optimizer).

I am not sure I see any potential issue, as long as the following setup sequence is always followed:
  - Instantiate datasets
  - Instantiate data loaders & task using datasets
  - Instantiate model using task
  - Instantiate optimizer using model
  - Instantiate scheduler using optimizer
  - ...

Is there a use case I'm not seeing?

Reply:
Given the original use case was to call `thelper.train.create_trainer` with the model and the custom optimizer, the approach seems fair.

Reply:
Thank you for the changes, I'll try it on my side right away.